staging/src/k8s.io/client-go/tools/portforward/portforward.go

```diff
@@ -391,7 +391,7 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 
 	go func() {
 		// inform server we're not sending any more data after copy unblocks
-		defer dataStream.Close()
```

Comment:
maybe we should keep the Close() here, send the upstream info to indicate local conn is closed.

And add the `dataStream.Reset()` after the 
```
select {
	case <-remoteDone:
	case <-localError:
	}
dataStream.Reset()
```
This way, when the local connection is closed we can inform the upstream server of the situation so it can have some time to clean up, and then we send the Reset

Reply:
yeah, I'm not sure we can close both directions of the stream in this goroutine... 

would it be better to make sure we reset the dataStream before we remove it, like this?

```diff
  defer pf.streamConn.RemoveStreams(dataStream)
+ defer dataStream.Reset()
```

Reply:
(rereading, I think my suggestion is the ~same as @aojea's but done in a defer so we do it even in panic cases)

Reply:
defer sounds better to me

Reply:
@nic-6443 can you please squash all commits and apply just the change Jordan says?

Reply:
@aojea @liggitt 
I tried the way proposed by Jordan and found it doesn't work because of the following code:
https://github.com/kubernetes/kubernetes/blob/8009d4d458530325fbbf1cf2afb81313597af804/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L411-L412
If we can't handle the blocking data in `dataStream` before this code, then `errorStream` will also be affected, causing `err = <-errorChan` to block.

Reply:
heh, this code is super tricky, thanks for catching it, do you mind squashing all commits?

Reply:
done.